src/routeGeneration/templateHelpers.ts

```diff
@@ -359,6 +363,95 @@ export class ValidationService {
     return new Buffer(value);
   }
 
+  public validateUnion(name: string, value: any, fieldErrors: FieldErrors, subSchemas: TsoaRoute.PropertySchema[] | undefined, parent = ''): any {
+    if (!subSchemas) { return; }
+
+    const subFieldErrors = subSchemas.map(subSchema => {
+      const subFieldErrors: FieldErrors = {};
+      this.ValidateParam(subSchema, value, name, subFieldErrors, parent);
+      return subFieldErrors;
+    });
+
+    if (!subFieldErrors.some(subFieldError => Object.keys(subFieldError).length === 0)) {
+      fieldErrors[parent + name] = {
+        message: `Could not match the union against any of the items. Issues: ${subFieldErrors}`,
+        value,
+      };
+      return;
+    }
+
+    return value;
+  }
+
+  public validateIntersection(name: string, value: any, fieldErrors: FieldErrors, subSchemas: TsoaRoute.PropertySchema[] | undefined, parent = ''): any {
+    if (!subSchemas) { return; }
+
+    const subFieldErrors = subSchemas.map(subSchema => {
+      const subFieldErrors: FieldErrors = {};
+      this.ValidateParam(subSchema, value, name, subFieldErrors, parent);
+      return subFieldErrors;
+    });
+
+    subFieldErrors.forEach(subFieldError => {
+      (Object as any).entries(subFieldError).forEach(([key, value]) => {
+        if (value.message.includes('excess')) {
+          delete subFieldError[key];
+        }
+      });
+    });
+
+    if (subFieldErrors.some(subFieldError => Object.keys(subFieldError).length !== 0)) {
+      fieldErrors[parent + name] = {
+        message: `Could not match the intersection against every type. Issues: ${subFieldErrors}`,
+        value,
+      };
+      return;
+    }
+
+    // Only Model definitions make sense here right now
+    const refNames = subSchemas.filter(subschema => subschema.ref).map(subschema => subschema.ref) as string[];
+    if (!refNames.every(refName => this.models[refName])) {
+      return value;
+    }
+
+    const reportedExcess = refNames.map(refName => this.models[refName]).reduce((acc, subSchema) => {
```

Comment:
`allowedProperties` can stay a `Set` because you're using the `Set.prototype.has` method (which is Big-O(1). But it's much slower to have `reportedExcess` be a `Set` since you never use any Big-O(1) operations on it. Additionally when you "unwrap" the Set later via the spread operator, you're iterating over the Set a second unnecessary time.

Please remove any conversions to `Set` where you aren't using `.has`

I've created a JsPerf demonstrating the advantages of not using Set in this circumstance:
https://jsperf.com/reduce-with-new-set-vs-seed/1

Reply:
The reason I am using a Set here is to avoid duplication here.

Reply:
If there is an actual excess property, every single Model in the intersection will report that.
But I think I can create only one Set at the end.